pkg/cmd/diff/diff.go

```diff
@@ -0,0 +1,159 @@
+
+	return cmd
+}
+
+func run(cmd *cobra.Command, datasource, rootID, old, new string, opts *option) error {
+	// human friendly normalization
+	rootID = strings.TrimSuffix(filepath.Base(rootID), ".json")
+	datasource = strings.TrimPrefix(datasource, "vuls-data-extracted-")
+
+	log.Printf("[INFO] Diff of %s, rootID: %s, commits: %.7s..%.7s", datasource, rootID, old, new)
+	wholeDiff, err := diff.Diff(datasource, rootID, old, new,
+		diff.WithOrg(opts.org), diff.WithDir(opts.dir), diff.WithFilter(opts.filter), diff.WithDiffAlg(opts.diffAlg))
+	if err != nil {
+		return errors.Wrapf(err, "get whole diff")
+	}
+
+	var writer = cmd.OutOrStdout()
+	if opts.outputDir != "" {
+		filename := fmt.Sprintf("diff_%s_%s_%.7s_%.7s.%s", datasource, rootID, old, new, func() string {
+			switch opts.format.String() {
+			case "text":
+				return "txt"
+			default:
+				return opts.format.String()
+			}
+		}())
+		f, err := os.Create(filepath.Join(opts.outputDir, filename))
+		if err != nil {
+			return errors.Wrapf(err, "open file. file: %s", filepath.Join(opts.outputDir, filename))
+		}
+		defer f.Close()
+		writer = f
+	}
+
+	err = output(wholeDiff, opts.format.String(), writer)
+	if err != nil {
+		return errors.Wrapf(err, "output")
+	}
+
+	return nil
+}
+
+func output(whole diff.WholeDiff, format string, writer io.Writer) error {
+	if format == "json" {
+		e := json.NewEncoder(writer)
+		e.SetIndent("", "\t")
+		if err := e.Encode(whole); err != nil {
+			return errors.Wrapf(err, "encode json")
+		}
+		return nil
+	}
+
+	fmt.Fprintf(writer, "root ID: %s\n", whole.RootID)
+	for n, e := range whole.Extracted {
+		fmt.Fprintf(writer, "\n======== EXTRACTED: %s ========\n", n)
+		fmt.Fprintf(writer, "  Compare URL: %s\n", e.CompareURL)
+		for _, c := range e.Files {
+			fmt.Fprintf(writer, "-------- COMPARISON: %s --------\n", c.Path)
+			fmt.Fprintf(writer, "  Old URL: %s\n", c.OldURL)
+			fmt.Fprintf(writer, "  New URL: %s\n", c.NewURL)
+			if c.OldPath != "" {
+				fmt.Fprintf(writer, "~~~~~~~~ DIFF [%s] %s %.7s..%.7s\n", n, c.Path+","+c.OldPath, e.Commits.Old, e.Commits.New)
+			} else {
+				fmt.Fprintf(writer, "~~~~~~~~ DIFF [%s] %s %.7s..%.7s\n", n, c.Path, e.Commits.Old, e.Commits.New)
+			}
+			fmt.Fprintf(writer, "%s\n", c.Diff)
+			fmt.Fprintf(writer, "~~~~~~~~ DIFF END\n")
+		}
+	}
+	for n, r := range whole.Raw {
+		fmt.Fprintf(writer, "\n======== RAW: %s ========\n", n)
+		fmt.Fprintf(writer, " Compare URL: %s\n", r.CompareURL)
+		for _, c := range r.Files {
+			fmt.Fprintf(writer, "-------- COMPARISON: %s --------\n", c.Path)
+			fmt.Fprintf(writer, "  Old URL: %s\n", c.OldURL)
+			fmt.Fprintf(writer, "  New URL: %s\n", c.NewURL)
+			var new = r.Commits.New
+			var old = r.Commits.Old
+			if new == "" {
+				new = "XXXXXXX"
+			}
+			if old == "" {
+				old = "XXXXXXX"
+			}
+			fmt.Fprintf(writer, "~~~~~~~~ DIFF [%s] %s %.7s..%.7s\n", n, c.Path, old, new)
+			fmt.Fprintf(writer, "%s\n", c.Diff)
+			fmt.Fprintf(writer, "~~~~~~~~ DIFF END\n")
+		}
+	}
```

Comment:
Do you think this will make it easier to read and adjust the display?
https://pkg.go.dev/text/template

Reply:
Tried but got not very good code. 🤷